server/etcdserver/v3_server.go

```diff
@@ -317,7 +317,32 @@ func (s *EtcdServer) LeaseRenew(ctx context.Context, id lease.LeaseID) (int64, e
 	return -1, errors.ErrCanceled
 }
 
-func (s *EtcdServer) LeaseTimeToLive(ctx context.Context, r *pb.LeaseTimeToLiveRequest) (*pb.LeaseTimeToLiveResponse, error) {
+func (s *EtcdServer) checkLeaseTimeToLive(ctx context.Context, leaseID lease.LeaseID) (error, uint64) {
+	rev := s.AuthStore().Revision()
+	if !s.AuthStore().IsAuthEnabled() {
+		return nil, rev
+	}
+	authInfo, err := s.AuthInfoFromCtx(ctx)
+	if err != nil {
+		return err, rev
+	}
+	if authInfo == nil {
+		return auth.ErrUserEmpty, rev
+	}
+
+	l := s.lessor.Lookup(leaseID)
+	if l != nil {
+		for _, key := range l.Keys() {
+			if err := s.AuthStore().IsRangePermitted(authInfo, []byte(key), []byte{}); err != nil {
+				return err, 0
+			}
+		}
+	}
```

Comment:
Before iterating over all the keys, I think we also need to check [LeaseTimeToLiveRequest.Keys](https://github.com/etcd-io/etcd/blob/8c5e9ad455e05a9435dad4a1a1e52733b538e3ba/api/etcdserverpb/rpc.pb.go#L2805); if it's `false`, it means users don't want to query al the keys at all, then no need to check the permission either on the server side.

We also need to add test to cover such case, such as the client has no range permission on a key, but it intentionally set `LeaseTimeToLiveRequest.Keys` to `false`, then it should not get an error response.

Reply:
Thanks for your comment, I'll update tomorrow.

Reply:
Updated, could you check @ahrtr ?